rust/src/shards/gui/widgets/plots/plot_bar.rs

```diff
@@ -0,0 +1,201 @@
+    (
+      cstr!("Horizontal"),
+      cstr!("Display the bars horizontally."),
+      BOOL_VAR_OR_NONE_SLICE,
+    )
+      .into(),
+    (
+      cstr!("Name"),
+      cstr!("Name of this chart, displayed in the plot legend."),
+      STRING_OR_NONE_SLICE,
+    )
+      .into(),
+  ];
+}
+
+impl Default for PlotBar {
+  fn default() -> Self {
+    let mut plot_ui = ParamVar::default();
+    plot_ui.set_name(PLOT_UI_NAME);
+    Self {
+      plot_ui,
+      requiring: Vec::new(),
+      color: ParamVar::default(),
+      bar_width: ParamVar::default(),
+      horizontal: ParamVar::default(),
+      name: ParamVar::default(),
+    }
+  }
+}
+
+impl Shard for PlotBar {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.PlotBar")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.PlotBar-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.PlotBar"
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &SEQ_OF_FLOAT2_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &SEQ_OF_FLOAT2_TYPES
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&BAR_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => Ok(self.color.set_param(value)),
+      1 => Ok(self.bar_width.set_param(value)),
+      2 => Ok(self.horizontal.set_param(value)),
+      3 => Ok(self.name.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.color.get_param(),
+      1 => self.bar_width.get_param(),
+      2 => self.horizontal.get_param(),
+      3 => self.name.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.PlotContext to the list of required variables
+    let exp_info = ExposedInfo {
+      exposedType: EGUI_PLOT_UI_TYPE,
+      name: self.plot_ui.get_name(),
+      help: cstr!("The exposed UI plot context.").into(),
+      ..ExposedInfo::default()
+    };
+    self.requiring.push(exp_info);
+
+    Some(&self.requiring)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.plot_ui.warmup(ctx);
+    self.color.warmup(ctx);
+    self.bar_width.warmup(ctx);
+    self.horizontal.warmup(ctx);
+    self.name.warmup(ctx);
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.name.cleanup();
+    self.horizontal.cleanup();
+    self.bar_width.cleanup();
+    self.color.cleanup();
+    self.plot_ui.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, _context: &Context, input: &Var) -> Result<Var, &str> {
+    let plot_ui: &mut egui::plot::PlotUi =
+      Var::from_object_ptr_mut_ref(*self.plot_ui.get(), &EGUI_PLOT_UI_TYPE)?;
+
+    let seq: Seq = input.try_into()?;
+    let points = seq.iter().map(|x| {
```

Comment:
hell a lot of allocations here, compared to imgui plot where I avoided allocations completely :(

Reply:
Can't be totally avoided, as `egui::plot::BarChart::new()` takes a `Vec<egui::plot::Bar>` as parameter.